### PR TITLE
Move hmac implementation to multiplatform library

### PR DESCRIPTION
**Pull request by tobiasheine**
### Description
This migrates the last hashing implementations from the java library to the multiplatform library.

### Considerations
Introduce `Hmac256Encoder` where the `JVM` implementation is using `javax.crypto` and the native one `platform.CoreCrypto`.

Migrate `Base16` to the multiplatform library using extension functions.

### Tests
The existing signer tests on the java-library, plus the added under `commonTest`, are passing using the `JVM` implementation.

To test the native implementation I introduced `iOSMain/HmacTestUtils`, which takes a `stringToSign` and `signingKey` and generates a signature using the native `Hmac256Encoder`.  The reason I did this is that the `Hmac256Encoder` is expecting a `KotlinByteArray` as input, which I would need to create from an `NSString` in the swift test. Again all this is just needed because the multiplatform gradle plugin misses support for running tests under `commonTest` on ios using a simulator.

### CODE REVIEW

File: ios/Sample/sampleTests/sampleTests.swift [hunk 1]
```diff
@@ -31,4 +31,16 @@ class sampleTests: XCTestCase {
         assert(contentHash == "5f1da1a2d0feb614dd03d71e87928b8e449ac87614479332aced3a701f916743")
     }
     
+    func testHmac256Signature() {
```
Comment (tobiasheine) [inline ios/Sample/sampleTests/sampleTests.swift:34]: This is the swift test double for https://github.com/novoda/aws-v4-signer-java/pull/6/files#diff-7b91a47366e48d7e31257e95ae72045d

File: java-library/src/main/java/uk/co/lucasweb/aws/v4/signer/Signer.java [hunk 2]
```diff
@@ -80,32 +78,16 @@ private static String buildAuthHeader(String accessKey, String credentialScope,
...
-        try {
-            String algorithm = HMAC_SHA256;
-            Mac mac = Mac.getInstance(algorithm);
-            SecretKeySpec signingKey = new SecretKeySpec(key, algorithm);
-            mac.init(signingKey);
-            return mac.doFinal(value.getBytes(UTF_8));
-        } catch (Exception e) {
-            throw new SigningException("Error signing request", e);
-        }
+        return Hmac256Encoder.INSTANCE.encode(key, value);
```
Comment (tobiasheine) [inline java-library/src/main/java/uk/co/lucasweb/aws/v4/signer/Signer.java:81]: This delegates to the `Hmac256Encoder` located in the multiplatform library.

File: kmp-library/src/iosMain/kotlin/com/novoda/aws/v4/signer/hash/Hmac256Encoder.kt [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package com.novoda.aws.v4.signer.hash
+
+import kotlinx.cinterop.addressOf
+import kotlinx.cinterop.convert
+import kotlinx.cinterop.usePinned
+import platform.CoreCrypto.CCHmac
+import platform.CoreCrypto.CC_SHA256_DIGEST_LENGTH
+import platform.CoreCrypto.kCCHmacAlgSHA256
+
+actual object Hmac256Encoder {
```
Comment (tobiasheine) [inline kmp-library/src/iosMain/kotlin/com/novoda/aws/v4/signer/hash/Hmac256Encoder.kt]: This is the native implementation using native c libraries under `platform.CoreCrypto` through [object pinning](https://blog.jetbrains.com/kotlin/2017/11/kotlinnative-v0-4-released-objective-c-interop-webassembly-and-more/).

File: kmp-library/src/jvmMain/kotlin/com/novoda/aws/v4/signer/hash/Hmac256Encoder.kt [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package com.novoda.aws.v4.signer.hash
+
+import javax.crypto.Mac
+import javax.crypto.spec.SecretKeySpec
+
+private const val HMAC_SHA256 = "HmacSHA256"
+
+actual object Hmac256Encoder {
```
Comment (tobiasheine) [inline kmp-library/src/jvmMain/kotlin/com/novoda/aws/v4/signer/hash/Hmac256Encoder.kt:8]: The `JVM` Hmac256Encoder implementation using `javax.crypto.Mac` as it was done previously in the `java-library`.

File: java-library/src/main/java/uk/co/lucasweb/aws/v4/signer/Signer.java (hunk 2 above)
Comment (juankysoriano) [reply java-library/src/main/java/uk/co/lucasweb/aws/v4/signer/Signer.java:81]: I can see that at the moment `Hmac256Encoder` is an `expected object`. The question is, when we have the implementation for `ios` and `jvm`, will `INSTANCE` resolve to the implementation for the platform that is being compiled?

File: kmp-library/src/jvmMain/kotlin/com/novoda/aws/v4/signer/hash/Hmac256Encoder.kt (hunk 4 above)
Comment (juankysoriano) [reply kmp-library/src/jvmMain/kotlin/com/novoda/aws/v4/signer/hash/Hmac256Encoder.kt:8]: :+1:

**Review by juankysoriano — APPROVED** [review]
LGTM (given my limited knowledge btw)

File: java-library/src/main/java/uk/co/lucasweb/aws/v4/signer/Signer.java (hunk 2 above)
Comment (tobiasheine) [reply java-library/src/main/java/uk/co/lucasweb/aws/v4/signer/Signer.java:81]: tldr; yes :)

The class we're looking at here is located in the `java-library` module, which [has a dependency](https://github.com/novoda/aws-v4-signer-java/commit/639b390201ade12d5e92e5d8dd830921cfc6f5a0) to the `kmp-library` module. The kotlin multiplatform gradle plugin makes sure that this is resolved to a dependency to the `jvmMain` source set of that library, which contains the `actual` jvm implementation of the `Hmac256Encoder`.

Once the java `Signer` class has been moved to the multiplatform library and migrated to kotlin, the `INSTANCE` prefix will be also gone.

**Pull request merged**